### PR TITLE
Prevent removal of aggregation nodes when grouping sets are present

### PR DESCRIPTION
**Pull request by ClarenceThreepwood**
## Description
Fixes #22171

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

### CODE REVIEW

**Review by feilong-liu — COMMENTED** [review]
LG.
For the unit tests, I think they need to be put in a separate test class corresponding to the optimizer itself.

**Review by ZacBlanco — APPROVED** [review]

**Review by ZacBlanco — APPROVED** [review]
Agree with Anant on moving the tests. Otherwise LGTM

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantAggregateDistinct.java [hunk 1]
```diff
@@ -45,7 +45,8 @@ public class RemoveRedundantAggregateDistinct
     private static boolean hasAggregations(AggregationNode node)
     {
         return ((GroupReference) node.getSource()).getLogicalProperties().isPresent() &&
-                !node.getAggregations().isEmpty();
+                !node.getAggregations().isEmpty() &&
+                node.getGroupingSetCount() == 1;
     }
```
[the remark this answers is not in the document]
Comment (ClarenceThreepwood) [reply presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantAggregateDistinct.java:51]: I think for logical equivalence, the second if will also need an else - i.e.
if (agg.getValue().isDistinct() ) {
                Set<VariableReferenceExpression> varAndGroupingKeySet =
                        Stream.concat(node.getGroupingKeys().stream().map(VariableReferenceExpression.class::cast),
                                        (agg.getValue()).getArguments().stream().map(VariableReferenceExpression.class::cast))
                                .collect(Collectors.toSet());
                if (((GroupReference) node.getSource()).getLogicalProperties().get().isDistinct(varAndGroupingKeySet)) {
                    aggregationsBuilder.put(agg.getKey(), removeDistinct(agg.getValue()));
                }
                else {
                     aggregationsBuilder.put(agg);
               }
            }
            else {
                aggregationsBuilder.put(agg);
            }


Also I modify this a bit anyway in #22013. Let's get back to this later if needed?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java [hunk 2]
```diff
@@ -1637,6 +1638,33 @@ public void testRedundantDistinctLimitNodeRemoval()
                                         tableScan("orders")))));
     }
 
+    @Test
+    public void testRedundantDistinctRemovalWithMultipleGroupingSets()
```
[the remark this answers is not in the document]
Comment (ClarenceThreepwood) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:1642]: For this test to give us the right signal, it needs to be run with more of the optimizer machinery than is available in TestRedundantDistinctRemoval (i.e. BaseRuleTest). In particular, for this test case to fail without the changes in this PR, the test would need to be run with RemoveRedundantDistinct (Rule), SimplifyPlanWithEmptyInput (PlanOptimizer), SimplifyRowExpressions. From what I can tell this is beyond the scope of BaseRuleTest and that is why the previous few tests in this file (testRedundantDistinctLimitNodeRemoval, testRedundantTopNNodeRemoval) live here. Open to suggestions if you have a better approach

[the remark this answers is not in the document]
Comment (aaneja) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:1642]: I see. We could modify the minimal plan optimizers that the Optimizer assert uses - https://github.com/prestodb/presto/blob/55113a37aecfee8fc018a9d8764d223cf800d696/presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/OptimizerAssert.java#L152

I'm fine with keeping these test here as-is though

[the remark this answers is not in the document]
Comment (aaneja) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:1642]: Can you add this test instead to `TestRedundantDistinctRemoval` - https://github.com/aaneja/presto/commit/e7dbce43c4abefede6fff5cf4353bc1e6e49c8d5

[the remark this answers is not in the document]
Comment (ClarenceThreepwood) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:1642]: I have added a simple test case to TestRedundantDistinctRemoval that triggers this situation using stats override as @aaneja suggested. I will leave the original test cases here for completeness

**Review by feilong-liu — APPROVED** [review]

**Review by aaneja — APPROVED** [review]

**Review by tdcmeehan — APPROVED** [review]

**Pull request merged**